File: Provenance/Controller/PVPCEControllerViewController.m
//
//  PVPCEControllerViewController.m
//  Provenance
//
//  Created by Joe Mattiello on 03/206/17.
//  Copyright © 2017 James Addyman. All rights reserved.
//

#import "PVPCEControllerViewController.h"
#import <PVMednafen/MednafenGameCore.h>

@interface PVPCEControllerViewController ()

@end

@implementation PVPCEControllerViewController

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    
    for (JSButton *button in [self.buttonGroup subviews])
    {
        if (![button isMemberOfClass:[JSButton class]])
        {
            continue; // skip over the PVButtonGroupOverlayView
        }
        
        if ([[[button titleLabel] text] isEqualToString:@"I"])
        {
            [button setTag:OEPCEButton1];
        }
        else if ([[[button titleLabel] text] isEqualToString:@"II"])
        {
            [button setTag:OEPCEButton2];
        }
        else if ([[[button titleLabel] text] isEqualToString:@"III"])
        {
            [button setTag:OEPCEButton3];
        }
        else if ([[[button titleLabel] text] isEqualToString:@"IV"])
        {
            [button setTag:OEPCEButton4];
        }
        else if ([[[button titleLabel] text] isEqualToString:@"V"])
        {
            [button setTag:OEPCEButton5];
        }
        else if ([[[button titleLabel] text] isEqualToString:@"VI"])
        {
            [button setTag:OEPCEButton6];
        }
    }
	
    [self.selectButton setTag:OEPCEButtonSelect];
    [self.startButton setTag:OEPCEButtonRun];
}

- (void)dPad:(JSDPad *)dPad didPressDirection:(JSDPadDirection)direction
{
    MednafenGameCore *pceCore = (MednafenGameCore *)self.emulatorCore;
    
    [pceCore didReleasePCEButton:OEPCEButtonUp forPlayer:0];
    [pceCore didReleasePCEButton:OEPCEButtonDown forPlayer:0];
    [pceCore didReleasePCEButton:OEPCEButtonLeft forPlayer:0];
    [pceCore didReleasePCEButton:OEPCEButtonRight forPlayer:0];
    
    switch (direction)
    {
        case JSDPadDirectionUpLeft:
            [pceCore didPushPCEButton:OEPCEButtonUp forPlayer:0];
            [pceCore didPushPCEButton:OEPCEButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionUp:
            [pceCore didPushPCEButton:OEPCEButtonUp forPlayer:0];
            break;
        case JSDPadDirectionUpRight:
            [pceCore didPushPCEButton:OEPCEButtonUp forPlayer:0];
            [pceCore didPushPCEButton:OEPCEButtonRight forPlayer:0];
            break;
        case JSDPadDirectionLeft:
            [pceCore didPushPCEButton:OEPCEButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionRight:
            [pceCore didPushPCEButton:OEPCEButtonRight forPlayer:0];
            break;
        case JSDPadDirectionDownLeft:
            [pceCore didPushPCEButton:OEPCEButtonDown forPlayer:0];
            [pceCore didPushPCEButton:OEPCEButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionDown:
            [pceCore didPushPCEButton:OEPCEButtonDown forPlayer:0];
            break;
        case JSDPadDirectionDownRight:
            [pceCore didPushPCEButton:OEPCEButtonDown forPlayer:0];
            [pceCore didPushPCEButton:OEPCEButtonRight forPlayer:0];
            break;
        default:
            break;
    }
    
    [self vibrate];
}

- (void)dPadDidReleaseDirection:(JSDPad *)dPad
{
    MednafenGameCore *pceCore = (MednafenGameCore *)self.emulatorCore;
    
    [pceCore didReleasePCEButton:OEPCEButtonUp forPlayer:0];
    [pceCore didReleasePCEButton:OEPCEButtonDown forPlayer:0];
    [pceCore didReleasePCEButton:OEPCEButtonLeft forPlayer:0];
    [pceCore didReleasePCEButton:OEPCEButtonRight forPlayer:0];
}

- (void)buttonPressed:(JSButton *)button
{
    MednafenGameCore *pceCore = (MednafenGameCore *)self.emulatorCore;
    [pceCore didPushPCEButton:[button tag] forPlayer:0];
    
    [self vibrate];
}

- (void)buttonReleased:(JSButton *)button
{
    MednafenGameCore *pceCore = (MednafenGameCore *)self.emulatorCore;
    [pceCore didReleasePCEButton:[button tag] forPlayer:0];
}

- (void)pressStartForPlayer:(NSUInteger)player
{
    MednafenGameCore *pceCore = (MednafenGameCore *)self.emulatorCore;
    [pceCore didPushPCEButton:OEPCEButtonMode forPlayer:player];
}

- (void)releaseStartForPlayer:(NSUInteger)player
{
    MednafenGameCore *pceCore = (MednafenGameCore *)self.emulatorCore;
    [pceCore didReleasePCEButton:OEPCEButtonMode forPlayer:player];
}

- (void)pressSelectForPlayer:(NSUInteger)player
{
    MednafenGameCore *pceCore = (MednafenGameCore *)self.emulatorCore;
    [pceCore didPushPCEButton:OEPCEButtonSelect forPlayer:player];
}

- (void)releaseSelectForPlayer:(NSUInteger)player
{
    MednafenGameCore *pceCore = (MednafenGameCore *)self.emulatorCore;
    [pceCore didReleasePCEButton:OEPCEButtonSelect forPlayer:player];
}

@end

File: Provenance/Controller/PVPCEControllerViewController.swift
//  Converted to Swift 4 by Swiftify v4.1.6640 - https://objectivec2swift.com/
//
//  PVPCEControllerViewController.swift
//  Provenance
//
//  Created by Joe Mattiello on 17/03/2018.
//  Copyright (c) 2018 Joe Mattiello. All rights reserved.
//

import PVSupport

fileprivate extension JSButton {
    var buttonTag: PVPCEButton {
        get {
            return PVPCEButton(rawValue: tag)!
        }
        set {
            tag = newValue.rawValue
        }
    }
}

class PVPCEControllerViewController: PVControllerViewController<PVPCESystemResponderClient> {

    override func layoutViews() {
        buttonGroup?.subviews.forEach {
            guard let button = $0 as? JSButton, let title = button.titleLabel?.text else {
                return
            }
            if title == "Ⅰ" {
                button.buttonTag = .button1
            } else if title == "Ⅱ" {
                button.buttonTag = .button2
            } else if title == "Ⅲ" {
                button.buttonTag = .button3
            } else if title == "Ⅳ" {
                button.buttonTag = .button4
            } else if title == "Ⅴ" {
                button.buttonTag = .button5
            } else if title == "Ⅵ" {
                button.buttonTag = .button6
            }
        }

        selectButton?.buttonTag = .select
        startButton?.buttonTag = .run
    }

    override func dPad(_ dPad: JSDPad, didPress direction: JSDPadDirection) {
        emulatorCore.didRelease(.up, forPlayer: 0)
        emulatorCore.didRelease(.down, forPlayer: 0)
        emulatorCore.didRelease(.left, forPlayer: 0)
        emulatorCore.didRelease(.right, forPlayer: 0)
        switch direction {
            case .upLeft:
                emulatorCore.didPush(.up, forPlayer: 0)
                emulatorCore.didPush(.left, forPlayer: 0)
            case .up:
                emulatorCore.didPush(.up, forPlayer: 0)
            case .upRight:
                emulatorCore.didPush(.up, forPlayer: 0)
                emulatorCore.didPush(.right, forPlayer: 0)
            case .left:
                emulatorCore.didPush(.left, forPlayer: 0)
            case .right:
                emulatorCore.didPush(.right, forPlayer: 0)
            case .downLeft:
                emulatorCore.didPush(.down, forPlayer: 0)
                emulatorCore.didPush(.left, forPlayer: 0)
            case .down:
                emulatorCore.didPush(.down, forPlayer: 0)
            case .downRight:
                emulatorCore.didPush(.down, forPlayer: 0)
                emulatorCore.didPush(.right, forPlayer: 0)
            default:
                break
        }
        vibrate()
    }

    override func dPadDidReleaseDirection(_ dPad: JSDPad) {
        emulatorCore.didRelease(.up, forPlayer: 0)
        emulatorCore.didRelease(.down, forPlayer: 0)
        emulatorCore.didRelease(.left, forPlayer: 0)
        emulatorCore.didRelease(.right, forPlayer: 0)
    }

    override func buttonPressed(_ button: JSButton) {
        emulatorCore.didPush(button.buttonTag, forPlayer: 0)
        vibrate()
    }

    override func buttonReleased(_ button: JSButton) {
        emulatorCore.didRelease(button.buttonTag, forPlayer: 0)
    }

    override func pressStart(forPlayer player: Int) {
        emulatorCore.didPush(.mode, forPlayer: player)
    }

    override func releaseStart(forPlayer player: Int) {
        emulatorCore.didRelease(.mode, forPlayer: player)
    }

    override func pressSelect(forPlayer player: Int) {
        emulatorCore.didPush(.select, forPlayer: player)
    }

    override func releaseSelect(forPlayer player: Int) {
        emulatorCore.didRelease(.select, forPlayer: player)
    }
}
